File: src/main/kotlin/marksync/services/Service.kt
package marksync.services

import marksync.Document
import marksync.uploader.FileInfo
import marksync.uploader.Uploader
import java.io.File

/**
 * Service class.
 */
abstract class Service(
    serviceName: String,
    val uploader: Uploader?
) {
    val metaFilename = "$METAFILE_PREFIX.$serviceName.yml"

    /**
     * Get all documents.
     *
     * @return Documents by keys
     */
    abstract fun getDocuments(): Map<String, ServiceDocument>

    /**
     * Get one document.
     *
     * @param id Document identifier
     * @return Document
     */
    abstract fun getDocument(id: String): ServiceDocument?

    /**
     * Get service document from Document.
     *
     * @param doc Document
     * @param dir Document directory
     * @return ServiceDocument object and expect digest
     */
    abstract fun toServiceDocument(doc: Document, dir: File): Pair<ServiceDocument, String?>?

    /**
     * Create meta data for service.
     *
     * @param dir Document directory
     */
    abstract fun createMeta(dir: File)

    /**
     * Save meta data to file.
     *
     * @param doc Document
     * @param dir Document directory
     * @param files Files
     */
    abstract fun saveMeta(doc: ServiceDocument, dir: File, files: ArrayList<FileInfo> = arrayListOf())

    /**
     * Update document to service.
     *
     * @param doc ServiceDocument object to update
     * @param message update message
     * @return Updated ServiceDocument object
     */
    abstract fun update(doc: ServiceDocument, message: String?): ServiceDocument?

    /**
     * Sync documents to service.
     *
     * @param dir Target directory
     * @param force Force update
     * @param message update message
     * @param checkOnly Check only
     * @param showDiff Show diff
     */
    fun sync(dir: File, force: Boolean, message: String?, checkOnly: Boolean, showDiff: Boolean) {
        val target = dir.path
        val doc = Document.of(dir)
        val (newDoc, expectDigest) = this.toServiceDocument(doc, dir) ?: return

        // check
        var doSync = false
        val docId = newDoc.getDocumentId()
        if (docId != null) {
            val oldDoc = getDocument(docId)
            if (oldDoc == null) {
                println("? $target: ($docId) not exists.")
            } else {
                val modifiedFiles = doc.files.filter { (filename, file) ->
                    !(newDoc.files.find { it.filename == filename }?.isIdenticalTo(file) ?: false)
                }.map { it.key }
                val modified = newDoc.isModified(oldDoc) || modifiedFiles.isNotEmpty()
                val drifted = expectDigest != null && oldDoc.getDigest() != expectDigest
                if (!modified && !drifted) {
                    println("  $target: not modified.")
                } else {
                    if (drifted) {
                        println("x $target: modified externally.")
                        println("  $expectDigest:${oldDoc.getDigest()}")
                    } else {
                        println("! $target")
                    }
                    if (!drifted || force) {
                        doSync = true
                    }
                    if (showDiff) {
                        newDoc.isModified(oldDoc, printDiff = true)
                        if (modifiedFiles.isNotEmpty()) {
                            println("  [files]")
                            modifiedFiles.forEach { println("  !$it") }
                        }
                        println("  ----------------------------------------------------------------")
                    }
                }
            }
        } else {
            println("+ $target")
            doSync = true
        }

        // sync
        if (doSync && !checkOnly) {
            if (uploader != null) {
                // upload files
                doc.files.forEach { (filename, file) ->
                    if (newDoc.files.find { it.filename == filename }?.isIdenticalTo(file) != true) {
                        uploader.upload(file)?.also { url ->
                            newDoc.files.find { it.filename == filename }?.let { newDoc.files.remove(it) }
                            newDoc.files.add(FileInfo(filename, file, url))
                            println("  ->uploaded. $filename to $url")
                        } ?: println("  ->upload failed. $filename")
                    }
                }
            }

            // update document
            update(newDoc, message)?.also { updatedDoc ->
                println("  ->updated. ${updatedDoc.getDocumentUrl()}")
                saveMeta(updatedDoc, dir, newDoc.files)
            } ?: println("  ->failed.")
        }
    }

    companion object {
        const val METAFILE_PREFIX = "marksync"
    }
}

File: src/marksync/remote/service.ts
import * as fs from "fs";
import {FileInfo} from "../uploader/file";
import {Uploader} from "../uploader/uploader";
import {LocalDocument} from "../document";
import {RemoteDocument} from "./document";

export abstract class RemoteService {
    static METAFILE_PREFIX = "marksync"

    protected readonly metaFilename: string;

    protected constructor(
        protected readonly serviceName: string,
        private readonly uploader: Uploader | null
    ) {
        this.metaFilename = `${RemoteService.METAFILE_PREFIX}.${serviceName}.yml`
    }

    /**
     * Get all documents.
     *
     * @return Documents by keys
     */
    abstract getDocuments(): Promise<{ [name: string]: RemoteDocument } | null>;

    /**
     * Get one document.
     *
     * @param id Document identifier
     * @return Document
     */
    abstract getDocument(id: string): Promise<RemoteDocument | null>;

    /**
     * Get service document from Document.
     *
     * @param doc Document
     * @param dirPath Document directory path
     * @return ServiceDocument object and expect digest
     */
    abstract toServiceDocument(doc: LocalDocument, dirPath: string): [RemoteDocument, string | null] | null;

    /**
     * Create meta data for service.
     *
     * @param dirPath Document directory path
     */
    abstract createMeta(dirPath: string): void;

    /**
     * Save meta data to file.
     *
     * @param doc Document
     * @param dirPath Document directory path
     * @param files Files
     */
    abstract saveMeta(doc: RemoteDocument, dirPath: string, files?: FileInfo[]): void;

    /**
     * Update document to service.
     *
     * @param doc ServiceDocument object to update
     * @param message update message
     * @return Updated ServiceDocument object
     */
    abstract update(doc: RemoteDocument, message: string | null): Promise<RemoteDocument | null>;

    /**
     * Prefetch documents.
     */
    async prefetch(): Promise<void> {
        await this.getDocuments();
    }

    /**
     * Sync documents to service.
     *
     * @param dirPath Target directory path
     * @param force Force update
     * @param message update message
     * @param checkOnly Check only
     * @param showDiff Show diff
     */
    async sync(dirPath: string, force: boolean, message: string | null, checkOnly: boolean, showDiff: boolean): Promise<void> {
        const target = dirPath;
        const newDocRes = this.toServiceDocument(LocalDocument.of(dirPath, this.serviceName), dirPath);
        if (!newDocRes) return;
        const [newDoc, expectDigest] = newDocRes;

        // check
        let doSync = false;
        const docId = newDoc.getDocumentId();
        if (docId !== null) {
            const oldDoc = await this.getDocument(docId);
            if (oldDoc === null) {
                process.stdout.write(`? ${target}: (${docId}) failed.\n`);
            } else {
                const modifiedFiles = Object.keys(newDoc.files).filter(filename => {
                    const file = newDoc.files[filename];
                    return fs.statSync(file).isFile()
                        && !(newDoc.fileInfoList.find(it => it.filename == filename)?.isIdenticalTo(file) || false);
                });
                const modified = await newDoc.isModified(oldDoc) || modifiedFiles.length > 0;
                const drifted = expectDigest !== null && await oldDoc.getDigest() !== expectDigest;
                if (!modified && !drifted) {
                    process.stdout.write(`  ${target}: not modified.\n`);
                } else {
                    if (drifted) {
                        process.stdout.write(`x ${target}: modified externally.\n`);
                        process.stdout.write(`  ${expectDigest}:${await oldDoc.getDigest()}\n`);
                    } else {
                        process.stdout.write(`! ${target}\n`);
                    }
                    if (!drifted || force) {
                        doSync = true
                    }
                    if (showDiff) {
                        await newDoc.isModified(oldDoc, true);
                        if (modifiedFiles.length > 0) {
                            process.stdout.write("  [files]\n");
                            modifiedFiles.forEach(it => {
                                process.stdout.write(`  !${it}\n`);
                            });
                        }
                        process.stdout.write("  ----------------------------------------------------------------\n");
                    }
                }
            }
        } else {
            process.stdout.write(`+ ${target}\n`);
            doSync = true;
        }

        // sync
        if (doSync && !checkOnly) {
            const newFileInfoList: FileInfo[] = [];
            for (const filename of Object.keys(newDoc.files)) {
                const filePath = newDoc.files[filename];
                const fileInfo = newDoc.fileInfoList.find(it => it.filename == filename);
                if (fs.statSync(filePath).isDirectory()) {
                    // link to another document
                    const anotherDocRes = this.toServiceDocument(LocalDocument.of(filePath, this.serviceName), filePath);
                    if (anotherDocRes !== null) {
                        const url = anotherDocRes[0].getDocumentUrl();
                        if (url !== null) {
                            newFileInfoList.push(new FileInfo({filename, url}));
                        }
                    }
                } else if (fs.statSync(filePath).isFile() && this.uploader !== null) {
                    if (fileInfo?.isIdenticalTo(filePath) != true) {
                        // upload file
                        const url = await this.uploader.upload(filePath);
                        if (url === null) {
                            process.stdout.write(`  ->upload failed. ${filename}\n`);
                            return;
                        }
                        newFileInfoList.push(FileInfo.of(filePath, url));
                        process.stdout.write(`  ->uploaded. ${filename} to ${url}\n`);
                    } else {
                        newFileInfoList.push(fileInfo);
                    }
                }
            }

            newDoc.fileInfoList = newFileInfoList;

            // update document
            const updatedDoc = await this.update(newDoc, message);
            if (updatedDoc !== null) {
                process.stdout.write(`  ->updated. ${updatedDoc.getDocumentUrl()}\n`);
                await this.saveMeta(updatedDoc, dirPath, newDoc.fileInfoList);
            } else {
                process.stdout.write("  ->failed.\n");
            }
        }
    }
}
